src/components/ConsoleInput.jsx

```diff
@@ -57,10 +62,65 @@ export default class ConsoleInput extends Component {
       editor.resize();
       editor.focus();
 
+      editor.commands.addCommand({
+        name: 'historyPrevious',
+        bindKey: 'Up',
+        exec: () => {
+          const {history, previousHistoryIndex} = this.props;
+          if (previousHistoryIndex >= history.size) {
+            return;
+          }
+
+          if (previousHistoryIndex === 0) {
+            onStoreCurrentConsoleInput(editor.getValue());
+          }
+          const newPreviousHistoryIndex = previousHistoryIndex + 1;
+
+          onSetPreviousHistoryIndex(newPreviousHistoryIndex);
+
+          const historyIndex = history.size - newPreviousHistoryIndex;
+          const {expression} = history.toList().get(historyIndex);
+          editor.setValue(expression);
+          editor.clearSelection();
```

Comment:
In general we want data flow in the application to work like this:

* Component receives props from Redux store, renders itself accordingly
* User does something, action is dispatched to Redux store
* Redux store updates its state based on action
* New props flow out to components from updated Redux state
* Components render themselves based on new props

In this case we are short-circuiting this process by having the editor update its internal state in the handler for the user action—thus the editor component’s UI state is no longer a function merely of props, but also of the internal updates that have been made to editor state based on user actions. This is only possible at all because this component uses an ACE editor instance, which provides a sort of imperative/mutable bubble inside the otherwise purely declarative React-Redux structure.

Since ultimately we want the component’s UI state to be purely a function of its props, I think we’ll need to have a concept of “current value of the console input” living in the Redux store. Then this component can receive that information as a prop, and call `setValue()` any time the received prop does not match the value currently held in the editor instance. This is precisely how the `<Editor>` component operates.

That information flow does impose certain additional requirements on the data shape in Redux, specifically that we need to store the current value of the console input, and update it every time the user adds or removes a character.

Although the proposed information flow probably feels like a lot of extra ceremony, my intuition is that it’ll actually make the code overall simpler—in particular I don’t think the event handlers in this component will need to do much heavy lifting, and the reducers and selectors can take on manageable chunks of logic which in combination produce the right effect.

A qualification that is in order: having “the component’s UI state…purely a function of its props” doesn’t strictly describe the status quo of the console in production today, since the value of the editor input is not described anywhere in props. This comes down to the choice between [controlled components](https://reactjs.org/docs/forms.html#controlled-components) vs. [uncontrolled](https://reactjs.org/docs/uncontrolled-components.html); my feeling is that controlled components are always preferable, but uncontrolled components are acceptable as long as the application code never needs to change the value of the input (a rule which is anyway broken [here](https://github.com/popcodeorg/popcode/blob/master/src/components/ConsoleInput.jsx#L63); I’m sorry!!)